Promises that overfulfill throw on overfulfillment

fuzzybinary opened the pull request:
This doesn't fix all the problems with overfulfillment but makes completing a zip or zipArray without all promises in the array providing results unlikely. I've left in checks for unfulfilled promises but those should no longer happen.

Commit:
Promises that overfulfill throw on overfulfillment

This doesn't fix all the problems with overfulfillment but makes completing a zip or zipArray without all promises in the array providing results unlikely. I've left in checks for unfulfilled promises but those should no longer happen.

kconner commented on Sources/Promise.swift:
```diff
@@ -233,11 +233,24 @@ public struct Promise<Value> {
      */
     public func inDispatchGroup(_ group: DispatchGroup) -> Promise<Value> {
+        var fulfilled = false;
+        let fulfillmentBarrier = DispatchQueue(label: "ThreadSafeResultsArray.queue", attributes: .concurrent)
```
These fields would need to be declared inside the worker block passed to the Promise returned from this function, or else they would be shared among multiple invocations of the Promise. In particular, `fulfilled` would be shared mutable state among concurrent operations which could then race. PinkyPromise Promises are not one-shot objects, they may be run multiple times with `.call`, so any extra tracked data like this should exist per call.

fuzzybinary replied:
Yes you're right, this should be one level up inside the promise.

kconner commented on Sources/Result.swift:
```diff
@@ -45,6 +45,19 @@ public extension Result where Failure == Error {
+            return .failure(PromiseError.overfulfilledZipPromise(atIndex: atIndex))
+        }
+        return self;
+    }
```
This is specific to Promise and I don't think it belongs in Result, which otherwise has no knowledge of Promise. Now that we are using Swift.Result, is there a generic `mapError` that Promise could call to do this specific work?

kconner commented on Sources/Promise.swift:
```diff
@@ -441,8 +454,16 @@ public func zipArray<T>(_ promises: [Promise<T>]) -> Promise<[T]> {
+            let unwrappedResults = results.enumerated().map { arg -> Result<T, Error> in
+                let (index, result) = arg
+                switch result {
+                case nil:
+                    return .failure(PromiseError.unfulfilledZipPromise(atIndex: index))
```
I think I remember seeing another PR that handled this situation smoothly by making this failing promise the default value of each result array item, and simplifying the type of `results` to make its items non-optional. Consider initializing `results` by mapping your range of indexes to failing Results.

fuzzybinary replied:
It does that higher up in the individual "zip" implementations. I agree that's a better implementation.

Commit:
Fix issues from code review.

Fix an issue where fulfillment parameters were in the wrong place. Move overfulfillment check out of Result. Rename fulfillment barrier.

kconner commented on Sources/Promise.swift:
```diff
@@ -234,10 +234,23 @@ public struct Promise<Value> {
     public func inDispatchGroup(_ group: DispatchGroup) -> Promise<Value> {
         return Promise { fulfill in
+            var fulfilled = false;
+            let fulfillmentBarrier = DispatchQueue(label: "ThreadSafeResultsArray.queue", attributes: .concurrent)
```
I think you're aware of the barrier queue in #80; is this meant to replace it or be an addition? If an addition, I think we could label each uniquely.

fuzzybinary replied:
Yeah, I'm aware of it. This is an addition I think, and I think they solve different problems. Whether or not #80 is required depends on how thread-safe writing to different points in an array is.

I will change the label though, good call.

kconner replied:
My hunch is that array underlying storage is theoretically fixed and therefore perhaps safe if the array is uniquely referenced, but all bets are off under COW conditions. But because I'm not sure, I accepted #80 as probably a good idea.

fuzzybinary commented:
@kconner Updated. Let me know if I addressed all your comments.

kconner commented on Sources/Promise.swift:
```diff
@@ -234,10 +234,23 @@ public struct Promise<Value> {
+                    // the promise to error out.
+                    guard !fulfilled else {
+                        let errorResult = Result<Value, Error>.failure(PromiseError.overfulfilledPromise)
+                        fulfill(errorResult)
```
In the case that you have already fulfilled, this is going to avoid calling `leave` a second time, but it will `fulfill` a second time, which breaks the promise contract (fulfill exactly once per invocation). Is it our position that, because that only happens when the wrapped promise breaks the contract, it's OK for us to do this in the name of expressing the problem safely?

kconner replied:
I think the alternative here is to log a warning and take no other action.

fuzzybinary replied:
I actually have that exact same question. This could cause some interesting errors where a result in an array gets modified out from under you potentially, but I'm not sure of the alternatives. 

My initial thought was to assert and then silently toss the second fulfillment in production. I can't imagine anyone is going to try to handle this error, but it could cause some weird silent failures if double fulfilment isn't caught in debug builds.

fuzzybinary replied:
Also, if we have that contract for all promises, should I be catching this at the call level instead?

kconner replied:
Well, we have asserted it as a _contract_ because up to now it has been the implementor's job to uphold :) That's why there was originally no safety in the library around these zip functions. PinkyPromise intends to have a small footprint.

#78 has an idea that is pretty compatible with this, which is a promise-wrapping method that adds a contract assertion, which makes it an opt-in feature for debugging.

kconner replied:
Ah, I think you must have used that one as a pattern in fact. There is a lot in common here.

kconner replied:
Personally, I think our priorities could be:

1. Prevent preventable crashes and our own mistakes regarding threads (such as in #75 and #80, and perhaps here re: `leave`)
2. Avoid doing extra bookkeeping work unless we are preventing crashes
3. Offer a way users can opt in to stronger checks (like #78)

I think that would mean:

1. In our `zip` functions, we do not need to add the extra layer of multiple fulfillment detection implicitly, but users can perform that wrapping on the items that they pass to `zip`.
2. In `inDispatchGroup`, we maybe should still prevent an additional call to `leave` even though we continue to call `fulfill` as many times as the wrapped promise fulfills.

fuzzybinary replied:
As I think about it, I think we probably don't have to assert the double fulfill in the general case as it should be easy to track down those issues. But in the zip / zipArray cases since it can cause some very hard to track down errors because we've made the assumption that the end-user is following the contract. That sounds like a perfect reason for an assert, and to avoid the double group.leave just ignore the second fulfill.

If we want to add something like #78 that could be useful, but something about adding an optional debugging feature feels weird to me.

kconner replied:
OK, are you suggesting we also skip the second fulfill here and otherwise leave the PR as is?

fuzzybinary replied:
Let me think about that. We should probably have a longer conversation in slack about it.

kconner commented on Sources/Promise.swift:
```diff
@@ -432,7 +445,9 @@ public func zipArray<T>(_ promises: [Promise<T>]) -> Promise<[T]> {
         let group = DispatchGroup()
 
-        var results: [Result<T, Error>?] = Array(repeating: nil, count: promises.count)
+        var results: [Result<T, Error>] = Array(0..<promises.count).lazy.map { (index) in
```
By the time you get to `.lazy`, you've already initialized the index array, so I'm not sure there are any savings to be found in `.lazy`. You should be able to map the range itself:

`(0..<promises.count).map…`

fuzzybinary replied:
Going off of https://lemire.me/blog/2018/01/23/initializing-arrays-quickly-in-swift-be-wary-of-seduns-initializers/ why using lazy is a good idea.

kconner replied:
Interesting! But the article's syntax is different; it makes the range lazy, not the array made from the range:

`Array(   (0..<1000).lazy.map { 100 * $0 }   )` (spaces for emphasis)

I would read that as meaning we need to do like:

`(0..<promises.count).lazy.map…` and if needed, wrap the whole thing with `Array( )` to convert from a lazy map sequence to the concrete type.

fuzzybinary replied:
Good catch. Fixed!

kconner commented on .swiftpm/xcode/package.xcworkspace/contents.xcworkspacedata:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Is this a file we should be .gitignoring? Should we ignore some parent path?

fuzzybinary replied:
Looks like we should ignore .swiftpm yeah

Commit:
Remove overCommit error.

Still prevent mismatched enter / leave on overfulfill but still allow overfulfillment.

Commit:
Merge branch 'develop' into bugfix/check-overfulfill

# Conflicts:
#	Sources/Promise.swift

fuzzybinary commented:
Going to close this and re-open with the new functionality.
